File: mcp4728.py
from typing import List, Optional, Tuple

import smbus


DAC_BITS: int = 12
MAX_CODE: int = int(2 ** DAC_BITS - 1)
INTERNAL_VREF: float = 2.048


class MCP4728:
    def __init__(self, Vdd: float, address: int = 0x60) -> None:
        self.Vdd: float = Vdd
        self.address = address
        self.bus = smbus.SMBus(1)

        self.max_voltage_internal = INTERNAL_VREF * (MAX_CODE / 2 ** DAC_BITS)
        self.max_voltage_external = Vdd * (MAX_CODE / 2 ** DAC_BITS)

        self.cur_voltages: List[float] = [0, 0, 0, 0]
        self.set_voltages(*self.cur_voltages)

    def _gain_vref_settings(self, v: float) -> Tuple[int, int, float]:
        """ Choose gain / reference settings for a given voltage, preferring
        the internal reference wherever possible since Vdd may be unreliable.

        Bit values:

        gain = 0: 1x gain
        gain = 1: 2x gain

        vref = 0: Vdd
        vref = 1: internal (2.048V)
        """

        assert v <= max(2 * self.max_voltage_internal, self.max_voltage_external)

        vref = 0
        gain = 0
        max_v = self.max_voltage_external

        if v <= 2 * self.max_voltage_internal:
            vref = 1
            max_v = self.max_voltage_internal

            if v > self.max_voltage_internal:
                gain = 1
                max_v *= 2

        return gain, vref, max_v

    def _get_update_bytes(self, channel: int, v: float) -> List[int]:
        assert channel < 4
        gain_bit, vref_bit, max_v = self._gain_vref_settings(v)
        v_int = int(MAX_CODE * v / max_v)

        b0 = 0b01000000 | (channel << 1)
        b1 = (vref_bit << 7) | (gain_bit << 4) | (v_int >> 8)
        b2 = (v_int & 0xff)

        return [b0, b1, b2]

    def set_voltages(
        self,
        v0: Optional[float] = None,
        v1: Optional[float] = None,
        v2: Optional[float] = None,
        v3: Optional[float] = None
    ) -> None:
        tgt_voltages: List[float] = [
            (v if v is not None else cur_v)
            for v, cur_v in zip([v0, v1, v2, v3], self.cur_voltages)
        ]

        to_update = [(i, v) for i, v in enumerate(tgt_voltages) if self.cur_voltages[i] != v]
        if len(to_update) < 1:
            return

        chan_data_blocks: List[List[int]] = [self._get_update_bytes(*iv) for iv in to_update]
        data: List[int] = [b for block in chan_data_blocks for b in block]

        self.bus.write_i2c_block_data(self.address, data[0], data[1:])
        self.cur_voltages = tgt_voltages

if __name__ == "__main__":
    import math
    import time

    mcp = MCP4728(Vdd=3.3)
    f = 262
    A = 3

    dac_range = (0.0, 3.3)
    out_range = (-3.3, 3.3)

    start = time.time()
    try:
        while True:
            t = time.time() - start
            v = A * math.sin(2 * math.pi * t * f)
            v = (v - out_range[0]) / (out_range[1] - out_range[0])
            v = v * (dac_range[1] - dac_range[0]) + dac_range[0]
            mcp.set_voltages(v)

            if t > 100:
                start = time.time()
    except KeyboardInterrupt:
        mcp.set_voltages(0)

File: src/mcp4728.cpp
#include "mcp4728.hpp"
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>
#include <unistd.h>
#include <stdexcept>
#include <cmath>
#include <algorithm>
#include <iostream>

MCP4728::MCP4728(double Vdd, uint8_t address, const std::string& i2c_device, bool auto_ref)
    : _Vdd(Vdd), _address(address), _cur_voltages(4, -1.0), _auto_ref(auto_ref) {
    // _cur_voltages is initialized to -1.0 to force first update.

    _i2c_fd = open(i2c_device.c_str(), O_RDWR);
    if (_i2c_fd < 0) {
        throw std::runtime_error("Failed to open I2C device: " + i2c_device);
    }

    if (ioctl(_i2c_fd, I2C_SLAVE, _address) < 0) {
        close(_i2c_fd);
        throw std::runtime_error("Failed to acquire bus access and/or talk to MCP4728.");
    }

    _max_voltage_internal = INTERNAL_VREF * (static_cast<double>(MAX_CODE) / (1 << DAC_BITS));
    _max_voltage_external = Vdd * (static_cast<double>(MAX_CODE) / (1 << DAC_BITS));

    // Initialize all channels to 0V
    set_voltages(0.0, 0.0, 0.0, 0.0);
}

MCP4728::~MCP4728() {
    if (_i2c_fd >= 0) {
        close(_i2c_fd);
    }
}

std::tuple<int, int, double> MCP4728::_gain_vref_settings(double v) {
    if (v > std::max(2.0 * _max_voltage_internal, _max_voltage_external)) {
        throw std::invalid_argument("Voltage " + std::to_string(v) + " exceeds maximum possible DAC output");
    }

    int vref = 0;
    int gain = 0;
    double max_v = _max_voltage_external;

    if (_auto_ref && v <= 2.0 * _max_voltage_internal) {
        vref = 1;
        max_v = _max_voltage_internal;

        if (v > _max_voltage_internal) {
            gain = 1;
            max_v *= 2.0;
        }
    }

    return {gain, vref, max_v};
}

std::vector<uint8_t> MCP4728::_get_update_bytes(int channel, double v) {
    const auto [gain_bit, vref_bit, max_v] = _gain_vref_settings(v);
    uint16_t v_int = static_cast<uint16_t>(MAX_CODE * v / max_v);
    if (v_int > MAX_CODE) {
        v_int = MAX_CODE;
    }

    // Command: 0100 0[DAC1][DAC0][UDAC_inv]
    //
    // Setting UDAC_inv and then later UDAC to 0 in byte_0 is fine because
    // subsequent byte_0s seem to be ignored in the multi-write command? The
    // first byte_0 has UDAC_inv = 0 which triggers the write.
    const uint8_t byte_0 = 0b01000000 | (channel << 1);
    const uint8_t byte_1 = (vref_bit << 7) | (gain_bit << 4) | (v_int >> 8);
    const uint8_t byte_2 = (v_int & 0xFF);

    return {byte_0, byte_1, byte_2};
}

void MCP4728::set_voltages(
    std::optional<double> v0,
    std::optional<double> v1,
    std::optional<double> v2,
    std::optional<double> v3
) {
    std::vector<std::optional<double>> inputs = {v0, v1, v2, v3};
    std::vector<double> tgt_voltages = _cur_voltages;
    std::vector<std::pair<int, double>> to_update;

    for (int i = 0; i < 4; ++i) {
        if (inputs[i].has_value()) {
            tgt_voltages[i] = inputs[i].value();
        }
        if (tgt_voltages[i] != _cur_voltages[i]) {
            to_update.push_back({i, tgt_voltages[i]});
        }
    }

    if (to_update.empty()) {
        return;
    }

    std::vector<uint8_t> data;
    for(const auto& [c, v] : to_update) {
        for(const auto& b : _get_update_bytes(c, v)) {
            data.push_back(b);
        }
    }

    if (write(_i2c_fd, data.data(), data.size()) != static_cast<ssize_t>(data.size())) {
        throw std::runtime_error("I2C write failed.");
    }

    _cur_voltages = tgt_voltages;
}
